src/segmentation.rs

```diff
@@ -31,99 +40,226 @@ pub fn segmentation_optimize<T: Pixel>(
       return;
     }
 
-    // A series of AWCY runs with deltas 13, 15, 17, 18, 19, 20, 21, 22, 23
-    // showed this to be the optimal one.
-    const TEMPORAL_RDO_QI_DELTA: i16 = 21;
-
     // Avoid going into lossless mode by never bringing qidx below 1.
     // Because base_q_idx changes more frequently than the segmentation
     // data, it is still possible for a segment to enter lossless, so
     // enforcement elsewhere is needed.
     let offset_lower_limit = 1 - fi.base_q_idx as i16;
 
-    // Fill in 3 slots with 0, delta, -delta. The slot IDs are also used in
-    // luma_chroma_mode_rdo() so if you change things here make sure to check
-    // that place too.
-    for i in 0..3 {
-      fs.segmentation.features[i][SegLvl::SEG_LVL_ALT_Q as usize] = true;
-      fs.segmentation.data[i][SegLvl::SEG_LVL_ALT_Q as usize] = match i {
-        0 => 0,
-        1 => TEMPORAL_RDO_QI_DELTA,
-        2 => (-TEMPORAL_RDO_QI_DELTA).max(offset_lower_limit),
-        _ => unreachable!(),
-      };
-    }
+    segmentation_optimize_inner(fi, fs, offset_lower_limit);
 
     /* Figure out parameters */
     fs.segmentation.preskip = false;
     fs.segmentation.last_active_segid = 0;
-    if fs.segmentation.enabled {
-      for i in 0..8 {
-        for j in 0..SegLvl::SEG_LVL_MAX as usize {
-          if fs.segmentation.features[i][j] {
-            fs.segmentation.last_active_segid = i as u8;
-            if j >= SegLvl::SEG_LVL_REF_FRAME as usize {
-              fs.segmentation.preskip = true;
-            }
+    for i in 0..MAX_SEGMENTS {
+      for j in 0..SegLvl::SEG_LVL_MAX as usize {
+        if fs.segmentation.features[i][j] {
+          fs.segmentation.last_active_segid = i as u8;
+          if j >= SegLvl::SEG_LVL_REF_FRAME as usize {
+            fs.segmentation.preskip = true;
           }
         }
       }
     }
   }
 }
 
+fn segmentation_optimize_inner<T: Pixel>(
+  fi: &FrameInvariants<T>, fs: &mut FrameState<T>, offset_lower_limit: i16,
+) {
+  const AVG_SEG: f64 = 3.0;
+
+  let coded_data = fi.coded_frame_data.as_ref().unwrap();
+  let segments = &coded_data.imp_b_segments;
+  let mut seg_counts = [0usize; MAX_SEGMENTS];
+  segments.iter().for_each(|&seg| {
+    // SAFETY: When selecting the segments, we ensured they are all within the range of 0-7
+    unsafe {
+      *seg_counts.get_unchecked_mut(seg as usize) += 1;
```

Comment:
I think the Rusty way to do this would be to have seg be a custom type with a setter than guarantees this. Not sure if you're up for a Type System Quest or not.

Reply:
I'm wondering how useful it would be outside of this function. If it can be reusable, I think it's worthwhile, I'm just struggling to wrap my head around that at the moment.

However, the modulus trick can be used here to eliminate the `unsafe` block at least.